src/libponyc/codegen/gentype.c

```diff
@@ -116,13 +111,41 @@ LLVMValueRef tbaa_metadata_for_box_type(compile_t* c, const char* box_name)
   return md->metadata;
 }
 
+void tbaa_tag(compile_t* c, LLVMValueRef metadata, LLVMValueRef instr)
+{
+  const char id[] = "tbaa";
+  unsigned md_kind = LLVMGetMDKindID(id, sizeof(id) - 1);
+
+  LLVMValueRef params[3];
+  params[0] = metadata;
+  params[1] = metadata;
+  params[2] = LLVMConstInt(c->i32, 0, false);
+
+  LLVMValueRef tag = LLVMMDNodeInContext(c->context, params, 3);
+  LLVMSetMetadata(instr, md_kind, tag);
+}
+
+void get_fieldinfo(ast_t* l_type, ast_t* right, ast_t** l_def,
+  ast_t** field, uint32_t* index)
+{
+  ast_t* d = (ast_t*)ast_data(l_type);
+  ast_t* f = ast_get(d, ast_name(right), NULL);
+  uint32_t i = (uint32_t)ast_index(f);
+
+  *l_def = d;
+  *field = f;
+  *index = i;
+}
+
 static LLVMValueRef make_tbaa_root(LLVMContextRef ctx)
 {
   const char str[] = "Pony TBAA";
   LLVMValueRef mdstr = LLVMMDStringInContext(ctx, str, sizeof(str) - 1);
   return LLVMMDNodeInContext(ctx, &mdstr, 1);
 }
 
+#if PONY_LLVM < 400
+
 static LLVMValueRef make_tbaa_descriptor(LLVMContextRef ctx, LLVMValueRef root)
```

Comment:
For LLVM 4 they need to be in the new path-aware format.  I had done some work on this but not finished it.  I'll take a further look.